// @histoire/plugin-vue@1.0.0-alpha.2/dist/bundled/client/server/run.js
import * as n from "virtual:$histoire-generated-global-setup";
import * as e from "virtual:$histoire-setup";
import { createApp as s, h as d } from "vue";
import f from "./Story.js";
import l from "./Variant.js";
async function h({ file: p, storyData: u, el: c }) {
  const { default: r } = await import(/* @vite-ignore */
    /* @vite-ignore */
    p.moduleId
  ), o = s({
    provide: {
      addStory(t) {
        u.push(t);
      }
    },
    render() {
      return d(r, {
        ref: "comp",
        data: p
      });
    }
  });
  o.component("Story", f), o.component("Variant", l);
  const a = {
    app: o,
    story: null,
    variant: null,
    addWrapper: () => {
    }
  };
  if (typeof (n == null ? void 0 : n.setupVue3) == "function") {
    const t = n.setupVue3;
    await t(a);
  }
  if (typeof (e == null ? void 0 : e.setupVue3) == "function") {
    const t = e.setupVue3;
    await t(a);
  }
  if (o.mount(c), r.doc) {
    const t = document.createElement("div");
    t.innerHTML = r.doc;
    const i = t.textContent;
    u.forEach((m) => {
      m.docsText = i;
    });
  }
  o.unmount();
}
export {
  h as run
};
